docs/proposals/post-deployment-hooks.md

```diff
@@ -0,0 +1,223 @@
+
+
+## Use cases
+
+1. As a Rails application developer, I want to perform a Rails database migration following an application deployment.
+2. As a Java application developer, I want to call a webhook following an application deployment.
+
+
+#### Use-case: Rails migrations
+
+New revisions of a Rails application often contain schema or other database updates which must accompany the new code deployment. Users should be able to specify a hook which performs a Rails migration as part of the application code deployment.
+
+Database migrations are complex and introduce downtime concerns. Here are [some examples](https://blog.rainforestqa.com/2014-06-27-zero-downtime-database-migrations) of zero-downtime Rails migration workflows.
+
+Deployments including database migrations must make special considerations:
+
+1.  Deployed code must be consistent with the currently deployed schema at any point in time.
+2.  Database or table locking must be minimized or eliminated to prevent service outages.
+
+The workflows which are effective at ensuring zero downtime migrations are typically multi-phased. For a user orchestrating a zero downtime migration deployment, it's likely the user needs to verify each deployment step discretely, with the option to abort and rollback after each phase.
+
+###### Container lifecycle hooks
+
+Container lifecycle hooks introduce problems with Rails migrations:
+
+1. The migration hook will execute once per pod in the deployment. This is a race and can result in concurrent migration attempts against a shared database. This burdens the hook author with coordinating the hooks or otherwise safeguarding the migration code to ensure it is safe for concurrent execution.
+  1. The uncoordinated aspect of these hooks is at also at odds with the general requirement to provide discrete step verification for the deployment.
+2. The migration hook will execute in the same pod as application containers, consuming resources allocated for the application.
+  1. This can cause instability, as it's unlikely the application pod resource allocation takes into account the temporarily increased requirements of a transient deployment step.
+
+###### Deployment hooks
+
+Deployment hooks satisfy this use case by providing a means to execute the hook only once per logical deployment. The hook is expressed as a run-once pod which provides the migration with its own resource allocation decoupled from the application.
+
+
+#### Use-case: Call a webhook
+
+Consider a Java application whose deployment should result in a webhook being called to notify some external system of the new deployment's availability.
+
+###### Container lifecycle hooks
+
+Container lifecycle hooks are inappropriate for this use case. The webhook should be called only once. Without a hook which coordinates with other replicated pod hook executions, the hook would:
```

Comment:
You can't guarantee only once - it's either at most once or at least.  Webhooks are probably not the right example, instead, make this be "must invoke a cloud api endpoint to notify it of the presence of new code".